paging/src/commonMain/kotlin/org/mobilenativefoundation/store/paging5/RealPager.kt

```diff
@@ -0,0 +1,158 @@
+        }
+
+        // Creating a child scope for coroutines.
+        val childScope = scope + Job()
+
+        // Launching a coroutine within the child scope for data loading.
+        childScope.launch {
+            // Locking the streams map to check and manage stream jobs.
+            mutexForAllStreams.withLock {
+                // Checking if there's no active stream for the key.
+                if (allStreams[key]?.isActive != true) {
+                    // Initializing the PagingData for the key.
+                    allPagingData[key] = emptyPagingData()
+
+                    val childrenKeys = mutableListOf<K>()
+
+                    // Main job to stream data for the key.
+                    val mainJob = launch {
+                        streamer(key).collect { response ->
+                            if (response is StoreReadResponse.Data<O>) {
+                                // Handling collection data response.
+                                (response as? StoreReadResponse.Data<StoreData.Collection<Id, SO>>)?.let { dataWithCollection ->
+
+                                    // Updating paging data and state flow inside a locked block for thread safety.
+                                    mutexForAllPagingData.withLock {
+                                        allPagingData[key] = pagingDataFrom(dataWithCollection.value.items)
+                                        val joinedData = joiner(allPagingData)
+                                        mutableStateFlow.value = joinedData
+                                    }
+
+                                    // For each item in the collection, initiate streaming and handling of single data.
+                                    dataWithCollection.value.items.forEach { single ->
+
+                                        val childKey = keyFactory.createFor(single.id)
+
+                                        (childKey as? K)?.let {
+                                            // Launching a coroutine for each single item.
+                                            val childJob = launch {
+                                                initStreamAndHandleSingle(single, childKey, key)
+                                            }
+
+                                            // Keeping track of child keys and jobs.
+                                            childrenKeys.add(childKey)
+                                            mutexForAllStreams.withLock {
+                                                allStreams[childKey] = childJob
+                                            }
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+
+                    // Storing the main job and handling its completion.
+                    allStreams[key] = mainJob
+
+                    mainJob.invokeOnCompletion {
```

Comment:
i think we shoiulda void child collectors. but if we cant, you can do this better by making their jobs a child of mainJob so that you won't need manual cleanup.